internal/services/notification/notification_test.go

```diff
@@ -0,0 +1,172 @@
+// Copyright 2023 Sorint.lab
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package notification
+
+import (
+	"context"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/rs/zerolog"
+	"github.com/sorintlab/errors"
+
+	"agola.io/agola/internal/services/config"
+	"agola.io/agola/internal/sqlg/sql"
+	"agola.io/agola/internal/testutil"
+	"agola.io/agola/services/notification/types"
+)
+
+func setupNotificationService(ctx context.Context, t *testing.T, log zerolog.Logger, dir string) *NotificationService {
+	dbType := testutil.DBType(t)
+	_, _, dbConnString := testutil.CreateDB(t, log, ctx, dir)
+
+	c := config.Config{
+		Notification: config.Notification{
+			DB: config.DB{
+				Type:       dbType,
+				ConnString: dbConnString,
+			},
+		},
+	}
+
+	rs, err := NewNotificationService(ctx, log, &c)
+	if err != nil {
+		t.Fatalf("unexpected err: %v", err)
+	}
+
+	return rs
+}
+
+func TestRunWebhookDelivery(t *testing.T) {
+	t.Parallel()
+
+	dir := t.TempDir()
+	ctx := context.Background()
+	log := testutil.NewLogger(t)
+
+	ns := setupNotificationService(ctx, t, log, dir)
+
+	t.Logf("starting ns")
+
+	time.Sleep(1 * time.Second)
+
+	// test delivery fail
+
+	runWebhook := createRunWebhook(t, ctx, ns)
+	runWebhookDelivery := createRunWebhookDelivery(t, ctx, ns, runWebhook.ID, types.NotDelivered)
+
+	if err := ns.handleRunWebhookDelivery(ctx, runWebhookDelivery.ID); err != nil {
+		t.Fatalf("unexpected err: %v", err)
+	}
+
+	runWebhookDeliverys := getRunWebhookDeliverys(t, ctx, ns)
+	if len(runWebhookDeliverys) != 1 {
+		t.Fatalf("expected %d runWebhook deliverys got: %d", 1, len(runWebhookDeliverys))
+	}
+	if runWebhookDeliverys[0].DeliveryStatus != types.DeliveryError {
+		t.Fatalf("expected runWebhook delivery status %q, got %q", types.DeliveryError, runWebhookDeliverys[0].DeliveryStatus)
+	}
+
+	// test delivery success
+
+	wr := setupWebhooksReceiver(ctx, t)
+	defer wr.stop()
+
+	t.Logf("starting webhooks client")
+
+	time.Sleep(1 * time.Second)
```

Comment:
This sleep is bad.

Reply:
I added sleep because the test was failing. I thought the webhookreceiver was not started and needed much time.
I try to remove

Reply:
webhookreceiver should start and return after is has been started.